### ISSUE TITLE
Type narrowing on FastifyReply

### ISSUE THREAD

**Issue by aadito123** [fastify/fastify#4755]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the issue has not already been raised

### Premise
Not really sure if this is a feature or a bug since methods on `Reply` that modify what the response schema should be, don't actually narrow at the type level when there is a broader type available.

### For example: 
```typescript
fastify.post('/',
    {
      schema: {
        response: {
          200: Type.Literal('Success!'), // strict string literal
          500: Type.String(), // broader string
        },
      },
    } as const,
    async (_, reply) => {
        // this should be a type error but it isn't
        reply.code(200).send('Hiya!');
    }
);
```

The reason it does not show a type error is because the type applied on the payload is `string | 'Success!` and since `string` is broader than `'Success!'`, it only applies `string`.

### Proposed Solution
The type can be narrowed by modifying generics in:
https://github.com/fastify/fastify/blob/d0d36964373b1c54d30a4e8de661da6ad186cc76/types/reply.d.ts#L34

to something like this:

```typescript
code<T extends keyof SchemaCompiler['response']>(statusCode: T): FastifyReply<RawServer, RawRequest, RawReply, RouteGeneric, ContextConfig, SchemaCompiler, TypeProvider, CallTypeProvider<TypeProvider, SchemaCompiler['response'][T]>>;
```

For consistency, this would need to be applied to the `status` method as well. 

Furthermore, some extra Typescript-ing would be needed to account for response schemas with keys like `2xx` but should be relatively simple to implement with string literals. 

I would be willing to make the pull request for this if this is something that seems useful.

**Comment by mcollina** [comment]
A PR would be amazing if it did not break any of the existing functionality!

**Comment by aadito123** [comment]
https://github.com/fastify/fastify/pull/4761#issuecomment-1554012146

Regarding the 'content-type' issue, maybe the schemas with validation for content types other than 'application/json' won't receive any of the type narrowing. This way there won't be any breaking changes and type narrowing can still be performed based on the code.

As for buffer and stream responses, generally those won't have a schema attached to them so I am not sure if I might be missing something.

**Comment by climba03003** [comment]
> As for buffer and stream responses, generally those won't have a schema attached to them so I am not sure if I might be missing something.

Schema is more than `type`, it also used to generate document.
Consider the below use-case.

```js
fastify.get('/multiple-content-type', {
  schema: {
    response: {
      200: {
        content: {
          'application/json': {
            schema: {
              type: 'object',
              // ...
            }
          },
          'application/xml': {
            schema: {
              type: 'string'
            }
          },
          'text/xml': {
            schema: {
              type: 'string'
            }
          }
        }

      }
    }
  }
}, function handler (request, reply) {
  const payload = { foo: 'bar' }
  // we narrow down the accept header here for demostration only
  switch(request.header['accept']) {
    case 'application/xml': {
      // using transform stream
      // JSONTOXMLTransform do not exist
      // just for demostration purpose
      const transform = new JSONTOXMLTransform()
      transform.push(payload)
      transform.push(null)
      return reply.code(200).send(transform)
    }
    case 'text/xml': {
      // use function to transform
      // jsonToXML do not exist
      // just for demostration purpose
      const xml = jsonToXML(payload)
      return reply.code(200).send(xml)
    }
    case 'application/json': 
    default: {
      return reply.code(200).send(payload)
    }
  }
})
```

**Comment by voxpelli** [comment]
Fixed by #4823, right?

**Comment by aadito123** [comment]
@voxpelli Yup. Closed with #4823
